OpenRA.Mods.Common/Traits/Health.cs

```diff
@@ -17,13 +19,40 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
 	{
 		[Desc("HitPoints")]
 		public readonly int HP = 0;
-
-		[Desc("Physical size of the unit used for damage calculations. Impacts within this radius apply full damage.")]
-		public readonly WDist Radius = new WDist(426);
-
 		[Desc("Trigger interfaces such as AnnounceOnKill?")]
 		public readonly bool NotifyAppliedDamage = true;
 
+		[FieldLoader.LoadUsing("LoadShape")]
+		public readonly IHitShape Shape;
+
+		static object LoadShape(MiniYaml yaml)
+		{
+			string shape = "Circle";
+			int radius = 426;
+			int2[] verts = null;
+
+			var shapeNode = yaml.Nodes.Find(n => n.Key == "Shape");
+			if (shapeNode != null)
+			{
+				shape = shapeNode.Value.Value;
+
+				var nodes = shapeNode.Value.ToDictionary();
+
+				if (nodes.ContainsKey("Radius"))
+					radius = FieldLoader.GetValue<int>("Radius", nodes["Radius"].Value);
+
+				if (nodes.ContainsKey("Vertices"))
+					verts = FieldLoader.GetValue<int2[]>("Vertices", nodes["Vertices"].Value);
+			}
+
+			switch (shape)
+			{
+				case "Rectangle": return new Rectangle(verts[0], verts[1]);
```

Comment:
Add `void Loaded()` to `IHitShape`:
```c#
	public interface IHitShape
	{
		WDist OuterRadius { get; }
		WDist DistanceFromEdge(WVec p);
		void DrawCombatOverlay(WorldRenderer wr, LineRenderer wlr, WPos center, WRot angle);
		void Loaded(); // Called once loaded from MiniYaml
	}
```

---
Modify hit shapes to something like:
```c#
	public class CircleHitShape : IHitShape
	{
		public WDist OuterRadius { get { return Radius; } }
		public readonly WDist Radius = new WDist(426);

		public void Loaded() { }

		...
	}
```
```c#
	public class CapsuleHitShape : IHitShape
	{
		public WDist OuterRadius { get { return outerRadius; } }
		[FieldLoader.required]
		public readonly int2 TopLeft;
		[FieldLoader.required]
		public readonly int2 BottomRight;
		public readonly WDist Radius = new WDist(426);
		WDist outerRadius;

		public void Loaded()
		{
			outerRadius = new WDist(radius + Math.Max(v.Length, w.Length));
		}

		...
	}
```
```c#
	public class RectangleHitShape : IHitShape
	{
		public WDist OuterRadius { get { return outerRadius; } }
		[FieldLoader.required]
		public readonly int2 Size;
		public readonly int2 Center = new int2(0, 0);
		public readonly WDist Radius = new WDist(426);
		WDist outerRadius;
		readonly WVec[] combatOverlayVerts;

		public void Loaded()
		{
			if (Size.X <= 0 || Size.Y <= 0)
				throw new YamlException("Size must be positive values");

			var tl = Center - Size / 2;
			var br = Center + Size / 2;
			combatOverlayVerts = new WVec[]
			{
				new WVec(tl.X, tl.Y, 0),
				new WVec(br.X, tl.Y, 0),
				new WVec(br.X, br.Y, 0),
				new WVec(tl.X, br.Y, 0),
				new WVec(tl.X, tl.Y, 0)
			};

			outerRadius = new WDist(Math.Max(tl.Length, br.Length));
		}

		...
	}
```
Note: each shape can have its own MiniYaml fields

---
Loader code:
```c#
		static object LoadShape(MiniYaml yaml)
		{
			var shapeNode = yaml.Nodes.Find(n => n.Key == "Shape");
			var shape = shapeNode == null || string.IsNullOrEmpty(shapeNode.Value) ? "Circle" : shapeNode.Value;
			var ret = Game.CreateObject<IHitShape>(shape + "HitShape");
			if (shapeNode != null)
			{
				try
				{
					FieldLoader.Load(ret, shapeNode);
				}
				catch (YamlException ex)
				{
					throw new YamlException("Shape " + shape + ": " + ex.Message));
				}
			}

			ret.Loaded();
			return ret;
		}
```